### PR TITLE
Problem142

### LINKED ISSUE #3024: Problem 142 (rust-side)
Rust part of this issue: https://github.com/nymtech/nym/issues/2999

### PR DESCRIPTION
**Pull request by jstuczyn**
# Description

Closes: https://github.com/nymtech/nym/issues/3024

Should potentially also address: https://github.com/nymtech/nym/issues/2782

So this PR slightly changes the behaviour of our vesting contract and thus please do not merge it until at least one, preferably two, other people reviewed it.

I've removed the split of delegated (and pledged) coins into "free" and "vesting". So whatever an account has put there, they can't be moved under any circumstances (unless, of course, they're undelegated).
As the result, `locked_coins` are now calculated as `vesting_coins - staked_coins` (as opposed to `vesting_coins - staked_vesting_coins`). This also updates `spendable_coins` that used `locked_coins` in calculation...

Furthermore, additional queries are exposed:
- `get_historical_vesting_staking_reward` => should return the total amount of rewards you have ever claimed into your vesting account (regardless of if you have withdrawn them into your liquid balance or not)
- `get_spendable_vested_coins` => for your `spendable_coins`, it should return how many of those come from the tokens that have vested. Note that vested coins are prioritised for withdrawals. It means that if for example you had 100nym spendable vested and 10nym spendable rewarded and decided to withdraw 5nym, it would result in you having now 95 spendable vested and unchanged 10nym spendable rewarded.
- `get_spendable_reward_coins` => for your `spendable_coins`, it should return how many of those come from the tokens that you got from the claimed rewards,
- `get_delegated_coins` => total amount of tokens you have delegated (via vesting)
- `get_pledged_coins` => total amount of tokens you have pledged (i.e. used for mixnode/gateway bonding)
- `get_staked_coins` => sum of your delegations and pledges

And obviously, those two are removed:
- `delegated_free`
- `delegated_vesting`

# Checklist:

- [ ] added a changelog entry to `CHANGELOG.md`

### CODE REVIEW

File: contracts/vesting/src/traits/vesting_account.rs [hunk 1]
```diff
@@ -58,35 +74,35 @@ pub trait VestingAccount {
...
-        storage: &dyn Storage,
-    ) -> Result<Coin, ContractError>;
-    /// See [/vesting-contract/struct.Account.html/method.get_pledged_vesting] for impl
-    fn get_pledged_vesting(
-        &self,
-        block_time: Option<Timestamp>,
-        env: &Env,
-        storage: &dyn Storage,
-    ) -> Result<Coin, ContractError>;
+    // fn get_delegated_free(
```
Comment (durch) [inline contracts/vesting/src/traits/vesting_account.rs]: I'd just straight up delete these

File: contracts/vesting/src/vesting/mod.rs [hunk 2]
```diff
@@ -929,497 +777,173 @@ mod tests {
...
-
-        let err = vesting_account.try_delegate_to_mixnode(
-            mix_id,
-            delegation.clone(),
-            &env,
-            &mut deps.storage,
-        );
-
-        assert!(err.is_err());
+        // let mut deps = init_contract();
```
Comment (durch) [inline contracts/vesting/src/vesting/mod.rs]: Same as above, kill it with fire

**Review by durch — APPROVED** [review]
Nice, love the `state_dumper` being commited as well.

File: contracts/vesting/src/vesting/mod.rs (hunk 2 above)
Comment (jstuczyn) [reply contracts/vesting/src/vesting/mod.rs]: :fire:

**Comment by merve64** [comment]
Hi. Did you see the this issue as well?  https://github.com/nymtech/nym/issues/3029 @jstuczyn This seems to be the same problem. There is total missing tokens: 385.645718 NYM
And they still can't be seen either in wallet nor in explorer. What to advice? still wait?

**Comment by jstuczyn** [comment]
@merve64 please check it now as we've just redeployed the vesting contract with this fix.

**Comment by merve64** [comment]
Hey! The user said that everything is fine now. @jstuczyn I am closing the issue.

**Pull request merged**